File: frontend/src/types/property.ts
export interface ValuationSummary {
  predicted_value: number | null;
  underval_score_pct: number | null;
  confidence_interval?: [number | null, number | null];
  feature_importances?: Record<string, number>;
  model_version: string | null;
  predicted_at?: string;
}

export interface PropertySummary {
  id: number;
  source: string;
  url: string;
  status: "for_sale" | "sold" | "withdrawn";
  property_type: "house" | "apartment" | "townhouse" | "land" | "rural" | null;
  address: string | null;
  address_suburb: string | null;
  address_postcode: string | null;
  suburb_id: number | null;
  latitude: number | null;
  longitude: number | null;
  bedrooms: number | null;
  bathrooms: number | null;
  car_spaces: number | null;
  land_size_sqm: number | null;
  floor_area_sqm: number | null;
  list_price: number | null;
  listed_at: string | null;
  valuation: ValuationSummary | null;
}

export interface PropertyDetail extends PropertySummary {
  address_street: string | null;
  year_built: number | null;
  price_guide_low: number | null;
  price_guide_high: number | null;
  sold_price: number | null;
  sold_at: string | null;
  description: string | null;
  features: string[] | null;
  images: string[];
  agent_name: string | null;
  agency_name: string | null;
  first_seen_at: string | null;
  last_seen_at: string | null;
}

export interface PropertyFilters {
  suburb?: string;
  postcode?: string;
  property_type?: string;
  bedrooms_min?: number;
  price_min?: number;
  price_max?: number;
  land_size_min?: number;
  underval_score_min?: number;
  bbox?: string;
  sort_by?: string;
  limit?: number;
  offset?: number;
}

File: backend/app/schemas/property.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class ValuationSummary(BaseModel):
    predicted_value: Optional[int] = None
    underval_score_pct: Optional[float] = None
    model_version: Optional[str] = None


class PropertySummary(BaseModel):
    id: int
    source: str
    url: str
    status: str
    property_type: Optional[str] = None
    address: Optional[str] = None
    address_suburb: Optional[str] = None
    address_postcode: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    car_spaces: Optional[int] = None
    land_size_sqm: Optional[float] = None
    list_price: Optional[int] = None  # AUD dollars
    listed_at: Optional[datetime] = None
    valuation: Optional[ValuationSummary] = None

    model_config = {"from_attributes": True}


class PropertyListResponse(BaseModel):
    items: list[dict]
    total: int
    limit: int
    offset: int


class PropertyDetail(BaseModel):
    id: int
    source: str
    url: str
    status: str
    property_type: Optional[str] = None
    address_street: Optional[str] = None
    address_suburb: Optional[str] = None
    address_postcode: Optional[str] = None
    suburb_id: Optional[int] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    land_size_sqm: Optional[float] = None
    floor_area_sqm: Optional[float] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    car_spaces: Optional[int] = None
    year_built: Optional[int] = None
    list_price: Optional[int] = None
    price_guide_low: Optional[int] = None
    price_guide_high: Optional[int] = None
    listed_at: Optional[datetime] = None
    sold_at: Optional[datetime] = None
    sold_price: Optional[int] = None
    description: Optional[str] = None
    features: Optional[list] = None
    agent_name: Optional[str] = None
    agency_name: Optional[str] = None
    images: Optional[list[str]] = None
    valuation: Optional[dict] = None

    model_config = {"from_attributes": True}
